File: context-engineering-intro/backend/app/core/sentry.py
"""
Sentry error monitoring and performance tracking integration.
"""
import os
import sentry_sdk
from sentry_sdk.integrations.fastapi import FastApiIntegration
from sentry_sdk.integrations.sqlalchemy import SqlalchemyIntegration
from sentry_sdk.integrations.redis import RedisIntegration
from sentry_sdk.integrations.celery import CeleryIntegration


def init_sentry():
    """
    Initialize Sentry error monitoring.

    Features:
    - Error tracking and alerting
    - Performance monitoring
    - Release tracking
    - Environment separation
    """
    sentry_dsn = os.getenv("SENTRY_DSN")

    if not sentry_dsn:
        print("⚠️  Sentry DSN not found - error monitoring disabled")
        return

    environment = os.getenv("SENTRY_ENVIRONMENT", os.getenv("ENVIRONMENT", "development"))
    traces_sample_rate = float(os.getenv("SENTRY_TRACES_SAMPLE_RATE", "1.0"))
    profiles_sample_rate = float(os.getenv("SENTRY_PROFILES_SAMPLE_RATE", "1.0"))

    sentry_sdk.init(
        dsn=sentry_dsn,
        environment=environment,

        # Performance Monitoring
        traces_sample_rate=traces_sample_rate,
        profiles_sample_rate=profiles_sample_rate,

        # Integrations
        integrations=[
            FastApiIntegration(transaction_style="url"),
            SqlalchemyIntegration(),
            RedisIntegration(),
            CeleryIntegration(),
        ],

        # Send PII (Personally Identifiable Information)
        send_default_pii=False,  # Set to True if you want to capture user data

        # Attach stack traces for pure messages
        attach_stacktrace=True,

        # Control what events are sent to Sentry
        before_send=before_send,

        # Release tracking (use git commit hash in production)
        # release=os.getenv("SENTRY_RELEASE", "eve-crm@1.0.0"),
    )

    print(f"✅ Sentry initialized - Environment: {environment}, Sample rate: {traces_sample_rate}")


def before_send(event, hint):
    """
    Filter or modify events before sending to Sentry.

    Use this to:
    - Remove sensitive data
    - Filter out certain errors
    - Add additional context
    """
    # Skip health check errors
    if event.get("transaction", "") == "/api/v1/health":
        return None

    # Filter out specific error types if needed
    if "exc_info" in hint:
        exc_type, exc_value, tb = hint["exc_info"]

        # Example: Don't send validation errors to Sentry
        # if isinstance(exc_value, ValidationError):
        #     return None

    return event


def capture_exception(error: Exception, **kwargs):
    """
    Manually capture an exception and send to Sentry.

    Usage:
        try:
            risky_operation()
        except Exception as e:
            capture_exception(e, extra={"user_id": 123})
            raise
    """
    sentry_sdk.capture_exception(error, **kwargs)


def capture_message(message: str, level: str = "info", **kwargs):
    """
    Capture a message (not an exception) and send to Sentry.

    Levels: "debug", "info", "warning", "error", "fatal"

    Usage:
        capture_message("Payment processed successfully", level="info", extra={"amount": 100})
    """
    sentry_sdk.capture_message(message, level=level, **kwargs)


def set_user(user_id: int = None, email: str = None, **kwargs):
    """
    Set user context for error tracking.

    Usage:
        set_user(user_id=123, email="user@example.com", role="admin")
    """
    sentry_sdk.set_user({
        "id": user_id,
        "email": email,
        **kwargs
    })


def set_context(name: str, context: dict):
    """
    Add custom context to error reports.

    Usage:
        set_context("payment", {"gateway": "stripe", "amount": 100})
    """
    sentry_sdk.set_context(name, context)


def add_breadcrumb(message: str, category: str = "default", level: str = "info", **kwargs):
    """
    Add a breadcrumb to track user actions leading to an error.

    Usage:
        add_breadcrumb("User clicked checkout button", category="user_action", level="info")
    """
    sentry_sdk.add_breadcrumb({
        "message": message,
        "category": category,
        "level": level,
        **kwargs
    })

File: context-engineering-intro/frontend/src/lib/sentry.ts
/**
 * Sentry Error Monitoring Configuration for Frontend
 *
 * This file initializes Sentry for error tracking and performance monitoring
 * in the Next.js frontend application.
 */

import * as Sentry from '@sentry/nextjs'

const SENTRY_DSN = process.env.NEXT_PUBLIC_SENTRY_DSN
const ENVIRONMENT = process.env.NEXT_PUBLIC_SENTRY_ENVIRONMENT || process.env.NEXT_PUBLIC_ENVIRONMENT || 'development'
const TRACES_SAMPLE_RATE = parseFloat(process.env.NEXT_PUBLIC_SENTRY_TRACES_SAMPLE_RATE || '1.0')

export function initSentry() {
  if (!SENTRY_DSN) {
    console.warn('⚠️  Sentry DSN not found - error monitoring disabled')
    return
  }

  Sentry.init({
    dsn: SENTRY_DSN,
    environment: ENVIRONMENT,

    // Performance Monitoring
    tracesSampleRate: TRACES_SAMPLE_RATE,

    // Session Replay (optional)
    replaysSessionSampleRate: 0.1, // 10% of sessions
    replaysOnErrorSampleRate: 1.0, // 100% of sessions with errors

    // Integrations
    integrations: [
      Sentry.replayIntegration({
        maskAllText: true,
        blockAllMedia: true,
      }),
    ],

    // Filter events before sending
    beforeSend(event, hint) {
      // Filter out development errors
      if (ENVIRONMENT === 'development') {
        console.log('Sentry Event (dev):', event)
        return null // Don't send in development
      }

      // Filter out certain errors
      if (event.exception) {
        const error = hint.originalException

        // Don't send network errors
        if (error instanceof Error && error.message.includes('NetworkError')) {
          return null
        }
      }

      return event
    },

    // Ignore certain errors
    ignoreErrors: [
      'ResizeObserver loop limit exceeded',
      'Non-Error promise rejection captured',
      'ChunkLoadError',
      'Loading chunk',
    ],
  })

  console.log(`✅ Sentry initialized - Environment: ${ENVIRONMENT}`)
}

/**
 * Capture an exception manually
 */
export function captureException(error: Error, context?: Record<string, any>) {
  Sentry.captureException(error, {
    extra: context,
  })
}

/**
 * Capture a message
 */
export function captureMessage(message: string, level: Sentry.SeverityLevel = 'info') {
  Sentry.captureMessage(message, level)
}

/**
 * Set user context for error tracking
 */
export function setUser(user: { id?: string; email?: string; username?: string } | null) {
  Sentry.setUser(user)
}

/**
 * Add breadcrumb for debugging
 */
export function addBreadcrumb(message: string, category?: string, data?: Record<string, any>) {
  Sentry.addBreadcrumb({
    message,
    category,
    data,
    level: 'info',
  })
}

/**
 * Set custom context
 */
export function setContext(name: string, context: Record<string, any>) {
  Sentry.setContext(name, context)
}
